### ISSUE TITLE
[FEATURE] Add nodeSelector as a parameter to VolumeSnapshotClass for backing images

### ISSUE THREAD

**Issue by dhedberg** [longhorn/longhorn#6526]
## Is your feature request related to a problem? Please describe (👍 if you like this request)

When creating a VolumeSnapshot using a class with `type: bi`, a backing image is created on some node in the cluster. It does not seem possible to control on which node(s) or disk(s) this initial image ends up.

Sometimes the user may know that the backing image is only going to be used on a particular set of hosts. Being able to directly create it on these hosts would (depending on the exact scenario) save on disk space, unnecessary network traffic and make creating new volumes from the snapshot faster.

## Describe the solution you'd like

Having nodeSelector as a parameter on VolumeSnapshotClasses, as is the case for StorageClasses, would improve the issue.

**Comment by innobead** [comment]
cc @ChanYiLin

**Comment by ChanYiLin** [comment]
VolumeSnapshot with type `bi` is actually implemented by creating backing image CR with type `export-from-volume`

I think we can make this feature request more general.
We can add node selector and disk selector to backing image CR.
No matter how this backing image is created (VolumeSnapshot, Create from UI), user has ability to control on which node/disk this backing image is prepared.

**Comment by innobead** [comment]
ref: #2856

**Comment by derekbit** [comment]
@ChanYiLin Can it move to `ready for testing`?

**Comment by ChanYiLin** [comment]
Hi @derekbit 
one PR left
the original request is to use csi
Here is the PR: https://github.com/longhorn/longhorn-manager/pull/2947

**Comment by chriscchien** [comment]
Hi @ChanYiLin,

`nodeSelector` and `diskSelector` set to backing image  successfully via csi VolumeSnapshot. 
But if multiple `nodeSelector` or `diskSelector` were set into backing image spec,  the backing image pod will not started, and file not start transfer to specific disk, could you take a look on this?

To reproduce. run [step 3](https://github.com/longhorn/longhorn/issues/6526#issuecomment-2089737138)

Backing image spec
```yaml
kl get backingimage -o yaml | grep spec -A25
  spec:
    checksum: ""
    diskFileSpecMap: {}
    disks: {}
    minNumberOfCopies: 3
    nodeSelector:
    - node1
    - node2
    secret: ""
    secretNamespace: ""
    sourceParameters:
      export-type: raw
      volume-name: test-vol1
    sourceType: export-from-volume
  status:
    checksum: ""
    diskFileStatusMap: null
    diskLastRefAtMap: null
    ownerID: ip-172-31-45-102
    size: 0
    uuid: 9288e62d
    virtualSize: 0
```

longhorn-manager logs
```
W0705 02:47:32.817829       1 logging.go:55] [core] [Channel #922 SubChannel #923]grpc: addrConn.createTransport failed to connect to {Addr: "10.42.2.28:8000", ServerName: "10.42.2.28:8000", }. Err: connection error: desc = "transport: Error while dialing: dial tcp 10.42.2.28:8000: connect: connection refused"
time="2024-07-05T02:47:32Z" level=warning msg="Failed to launch a gRPC client during cleanup, will skip deleting the file for backing image snapshot-305dcc9d-c92a-47da-989-84f9bb1f(7d633cf7)" func="controller.(*BackingImageManagerController).cleanupBackingImageManager" file="backing_image_manager_controller.go:357" backingImageManager=backing-image-manager-b2c6-7d9d controller=longhorn-backing-image-manager diskUUID=7d9d4c57-0f7c-47fc-a0d2-f8eac35dac79 error="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial tcp 10.42.2.28:8000: connect: connection refused\"" node=ip-172-31-32-254 nodeID=ip-172-31-32-254
time="2024-07-05T02:47:33Z" level=warning msg="Failed to find backing image manager for pod, may be deleted" func="controller.(*BackingImageManagerController).enqueueForBackingImageManagerPod" file="backing_image_manager_controller.go:1065" controller=longhorn-backing-image-manager node=ip-172-31-32-254 pod=backing-image-manager-b2c6-7d9d
time="2024-07-05T02:47:33Z" level=warning msg="Failed to find backing image manager for pod, may be deleted" func="controller.(*BackingImageManagerController).enqueueForBackingImageManagerPod" file="backing_image_manager_controller.go:1065" controller=longhorn-backing-image-manager node=ip-172-31-32-254 pod=backing-image-manager-b2c6-7d9d
```

**Comment by ChanYiLin** [comment]
About the tags:

The tags here is for setting the characteristic for a set of disks
for example: 
- users may label all the ssd disks with the tags: ["ssd"]
- users may label all the fast disks with the tags: ["fast"]
so there are three set of disks
1. tags: []
2. tags: ["ssd"]
3. tags: ["ssd", "fast"]

so if users want to schedule the volume to such ssd and fast disks, they should add the selectors: ["ssd", "fast"], then the disks will only be scheduled to `set3`
But if users only want to schedule to the ssd volume and don't care if it is slow, they should add the selectors: ["ssd"], then the disks will be scheduled to `set2` and `set3`

> When multiple tags are specified for a volume, the disk and the node (the disk belong to) must have all the specified tags to become usable.

In other words, the selectors of a volume **need to be a subset of** the tags of the nodes/disks can it be scheduled on that nodes/disks

**Comment by ChanYiLin** [comment]
Hi @chriscchien 
you can change the tags and selectors to test it again.

**Comment by chriscchien** [comment]
Verified **pass** on longhorn master(longhorn-manager `4aa94f`) with test [steps](https://github.com/longhorn/longhorn/issues/6526#issuecomment-2089737138)

`nodeSelector` and `diskSelector` set to backing image successfully via csi VolumeSnapshot and backing image file followed the selectors [rules](https://longhorn.io/docs/1.6.2/nodes-and-volumes/nodes/storage-tags/#usage) to transfer.

**Comment by derekbit** [comment]
Doc is updated in https://github.com/longhorn/website/pull/904
